src/vote_graph.rs

```diff
@@ -87,17 +87,17 @@ pub struct VoteGraph<H: Hash + Eq, N, V> {
 
 impl<H, N, V> VoteGraph<H, N, V> where
 	H: Hash + Eq + Clone + Ord + Debug,
-	V: AddAssign + Default + Clone + Debug,
+	V: for<'a> AddAssign<&'a V> + Default + Clone + Debug,
 	N: Copy + Debug + BlockNumberOps,
 {
 	/// Create a new `VoteGraph` with base node as given.
-	pub fn new(base_hash: H, base_number: N) -> Self {
+	pub fn new(base_hash: H, base_number: N, base_node: V) -> Self {
```

Comment:
do we have any use-case where we are initializing this to non-default? The `Default` ~ `Zero` property is important for a lot of the GHOST-stuff and this may discourage that.

Reply:
This was primarily motivated by the fact that making use of a type parameter constraint (i.e. trait bound) in a method that whereby the type parameter appears nowhere in the method's arguments necessarily leaves the parameter uninstantiated in expressions of the form
```rust
let graph = VoteGraph::new(...);
```
and thus makes type inference harder. Previously `V` could then be instantiated by the type system through observing subsequent occurrences of `graph.insert(...)`. Now the parameter of `insert` is some `W` such that `V: for<'a> AddAssign<&'a W>`, so `V` is still undetermined. It can of course be solved by giving an explicit type ascription to expressions such as the above (`let graph: ... = `), it just seemed to me that passing the base node in the constructor is preferable as it is more explicit and avoids this type ambiguity.

Generally the `VoteGraph` has no clue what `V::default()` produces, i.e. there was and is no `Default` ~ `Zero` property. However, it is certainly true that this allows constructing a `VoteGraph` with a base node that is different from what the default instance returns and the latter is used inside the `VoteGraph` to construct intermediate new nodes on demand. If that is undesirable, I can certainly go with the extra type annotations.